src/stdhasher.cr

```diff
@@ -0,0 +1,248 @@
+require "secure_random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+#   class MyHasher
+#     def <<(v : Int::Primitive)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(b : Bytes)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(n : Nil)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(v)
+#       v.hash(self)
+#       self
+#     end
+#
+#     # digest returns hashsum for current state without state mutation
+#     # Note: hashsum should implement commutative `+` if MyHasher used
+#     # for `Hasher#hash(hasher)`
+#     def digest
+#     end
+#
+#     def clone_build
+#       with_state_copy do |copy|
+#         yield copy
+#         copy.digest
+#       end
+#     end
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+#
+# Also it has specialized methods for primitive keys with different seeds.
+struct StdHasher
+  @@seed = StaticArray(UInt32, 7).new{|i| 0_u32}
+  def self.init
+    raise "#{@@seed[0]} #{@@seed[1]}" unless @@seed[0] == 0_u32 && @@seed[1] == 0_u32
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+  end
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
```

Comment:
It still will be slow for heavy weight hash functions.
And will be unoptimal for suggested hasher because of weird LLVM optimization passes (need to be rechecked, though).

But if you really want "hash(hasher) have to return hasher" to be part of protocol (for safety), I will agree.

Note, generic hasher '<<' function then should look like:
```crystal
struct MyHasher
  def <<(v)
    self.copy(v.hash(self))
    self
  end
end
```
will it work?